modules/packages/LinearAlgebra.chpl

```diff
@@ -605,6 +604,60 @@ private inline proc _raw(D: domain(1), i) {
 // Matrix Structure
 //
 
+
+/*
+   Return a Vector containing the diagonal elements of ``A`` if the argument ``A`` is of rank 2.
+   Return a diagonal Matrix whose diagonal contains elements of ``A`` if argument ``A`` is of rank 1.
+ */
+proc diag(A: [?Adom] ?eltType){
+  if(Adom.rank == 2) then{
+    if(!isSquare(A)) then halt("diag supports only square matrices");
+    return _diag_vec(A);  
+  }
+  else if(Adom.rank == 1) then {
+    return _diag_mat(A);
+  }
+  else{
+    compilerError("A must have rank 2 or less");
+  }
+
+}
+
+/*
+  Return a Vector containing the off diagonal elements of ``A``
+*/
+proc offdiag(A:[?Adom] ?eltType){
+  if(Adom.rank != 2) then halt("offdiag expects argument to be of rank 2");
+  if(!isSquare(A)) then halt ("offdiag supports only square matrices");
+  return _diag_vec(A,trans=true);
+}
+
+proc _diag_vec(A:[?Adom] ?eltType, trans=false){
```

Comment:
This may be moot if we use the definition of off-diagonal I gave above, but I suggest using ``trans`` in a linear algebra setting only when a matrix transpose is happening, which is not the case here. It just is a little confusing otherwise...